docs/datastorage/codecs.md

```diff
@@ -0,0 +1,437 @@
+
+A => B | B => A | Transform Method
+:---:  | :---:  | :---
+Yes    | Yes    | `#xmap`
+Yes    | No     | `#flatComapMap`
+No     | Yes    | `#comapFlatMap`
+No     | No     | `#flatXMap`
+
+```java
+// Given an string codec to convert to a integer
+// Not all strings can become integers (A !=> B)
+// All integers can become strings (B => A)
+public static final Codec<Integer> INT_CODEC = Codec.STRING.comapFlatMap(
+  s -> { // Return data result containing error on failure
+    try {
+      return DataResult.success(Integer.valueOf(s));
+    } catch (NumberFormatException e) {
+      return DataResult.error(s + " is not an integer.");
+    }
+  },
+  Integer::toString // Regular function
+);
+```
+
+```js
+// Will return 5
+"5"
+
+// Will error, not an integer
+"value"
+```
+
+#### Range Codecs
+
+Range codecs are an implementation of `#flatXMap` which returns an error `DataResult` if the value is not inclusively between the set minimum and maximum. The value is still provided as a partial result if outside the bounds. There are implementations for integers, floats, and doubles via `#intRange`, `#floatRange`, and `#doubleRange` respectively.
+
+```java
+public static final Codec<Integer> RANGE_CODEC = Codec.intRange(0, 4); 
+```
+
+```js
+// Will be valid, inside [0, 4]
+4
+
+// Will error, outside [0, 4]
+5
+```
+
+### Defaults
+
+If the result of encoding or decoding fails, a default value can be supplied instead via `Codec#orElse` or `Codec#orElseGet`.
+
+```java
+public static final Codec<Integer> DEFAULT_CODEC = Codec.INT.orElse(0); // Can also be a supplied value via #orElseGet
+```
+
+```js
+// Not an integer, defaults to 0
+"value"
+```
+
+### Unit
+
+A codec which supplies an in-code value and encodes to nothing can be represented using `Codec#unit`. This is useful if a codec uses a non-encodable entry within the data object.
+
+```java
+public static final Codec<IForgeRegistry<Block>> UNIT_CODEC = Codec.unit(
+  () -> ForgeRegistries.BLOCKS // Can also be a raw value
+);
+```
+
+```js
+// Nothing here, will return block registry codec
+```
+
+### List
+
+A codec for a list of objects can be generated from an object codec via `Codec#listOf`.
+
+```java
+// BlockPos#CODEC is a Codec<BlockPos>
+public static final Codec<List<BlockPos>> LIST_CODEC = BlockPos.CODEC.listOf();
+```
+
+```js
+// Encoded List<BlockPos>
+[
+  [1, 2, 3], // BlockPos(1, 2, 3)
+  [4, 5, 6], // BlockPos(4, 5, 6)
+  [7, 8, 9]  // BlockPos(7, 8, 9)
+]
+```
+
+List objects decoded using a list codec are stored in an **immutable** list. If a mutable list is needed, a [transformer] should be applied to the list codec.
+
+### Map
+
+A codec for a map of keys and value objects can be generated from two codecs via `Codec#unboundedMap`.
```

Comment:
This only mentions `#unboundedMap` which is a map with an unbounded set of keys, but not `#simpleMap` which is a map for a bounded set of keys (as provided by a `Keyable`). This should either mention both, or at least clarify what an "unbounded map" means.

Reply:
`#simpleMap` is a bit complicated for users to understand since it compresses a map into a list where the number of keys are fixed and ordered. This is one of the few codecs I would rather avoid mentioning since it is likely most modders will never use it nor really understand it.